### PR TITLE
Add a utility to help with S3 object renaming

### PR DESCRIPTION
**Pull request by ian-noaa**
Add a script to do bulk renaming in S3. The script updates keys with `YYYY-MM-DDTHH` timestamps in them to `YYYY-MM-DDTHH:MM` so that our timestamps are standardized.

Additionally, add a handy Poetry entrypoint so that this can be run with `poetry run`.

### CODE REVIEW

File: src/utils/s3/s3_bulk_renaming.py
Comment (esheehan-gsl) [inline src/utils/s3/s3_bulk_renaming.py:43]: Considering that this regex is used again in `update_timestamp`, I’d consider pulling it out and pre-compiling it at the top of the module into a global variable between your imports and the first function definition.

File: tests/utils/s3_bulk_renaming_test.py
Comment (esheehan-gsl) [inline tests/utils/s3_bulk_renaming_test.py:94]: Since these are all effectively the same test, I’d just write one test function and parametrize it:

```suggestion
@pytest.mark.parametrize(
    "desired,expected",
    (
        (
            (
                "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
                "2023-06-06T16/ges/observation/1.1"
            ),
            (
                "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
                "2023-06-06T16:00/ges/observation/1.1"
            ),
        ),
        ("foo/2023-06-06T16/bar", "foo/2023-06-06T16:00/bar"),
        ("2023-06-06T16", "2023-06-06T16:00"),
        ("2023-06-06T16:00", "2023-06-06T16:00"),
        ("foo/bar", "foo/bar"),
    )
)
def test_update_timestamp(desired, expected):
    assert desired == update_timestamp(key)
```

File: tests/utils/s3_bulk_renaming_test.py
Comment (esheehan-gsl) [inline tests/utils/s3_bulk_renaming_test.py:56]: Good catch. I was wondering why that function tested for both timestamp formats with and without minutes.

File: src/utils/s3/s3_bulk_renaming.py
Comment (esheehan-gsl) [inline src/utils/s3/s3_bulk_renaming.py:25]: I think I would suggest letting this exception bubble out of `list_objects` and put your try-except block with `sys.exit` in `main()`.

One reason to break up these scripts into functions instead of just one imperative script that executes start-to-finish is that it makes it possible to use the module as a library or to execute it directly. This is a pattern we see in the Python standard library with modules like `http` where you can import it to build an HTTP server or run it on the commandline to start up a simple HTTP server.

So if we think about using this module as a library, catching the `NoCredentialsError` and calling `sys.exit` could really interfere with someone’s ability to use this function in another program.

**Review by esheehan-gsl — COMMENTED** [review]
These are just a couple of small suggestions that I don’t think need to be implemented before merging. It’s up to you if you want to accept or reject the suggestions here.

**Review by esheehan-gsl — DISMISSED** [review]
I know this is a real nitpick, but all of the other test files are `test_*.py`, but this one is `*_test.py`.

I can definitely see an argument for using the test suffix instead of the prefix the way I have been, and I have no objection to making that change. But if we do that, I think the way I’d like to do it is to name this one with the prefix the way the other test files are named, and then open a new PR that just renames everything to use the suffix instead.

**Comment by ian-noaa** [comment]
Oh that's a good catch. I was recently doing some testing in Go and am pretty sure that's a leftover from doing things the "Go way". The `*_test.go` ending is handled specially there and I mindlessly copied that convention here.

Personally, I'm okay with the prefix like you've been using. As long as it's consistent!

File: tests/utils/s3_bulk_renaming_test.py
Comment (ian-noaa) [reply tests/utils/s3_bulk_renaming_test.py]: Ah, that's a nice improvement! This test always felt a bit cludgy to me - I haven't played with parameterized tests too much yet but this is a great usecase.

File: src/utils/s3/s3_bulk_renaming.py
Comment (ian-noaa) [reply src/utils/s3/s3_bulk_renaming.py:25]: That's a great point. Thanks for the suggestion!

File: src/utils/s3/s3_bulk_renaming.py
Comment (ian-noaa) [reply src/utils/s3/s3_bulk_renaming.py]: Good point. Do you know if there are naming conventions for precompiled regexes?

**Comment by ian-noaa** [comment]
@esheehan-gsl - thanks for the review! I ended up applying your suggestions and updating the test file name. Unfortunately, GitHub was blocking the merge until you're able to re-review this PR. Since you're out at the moment, I went ahead and "dismissed" your review so I can merge.

**Pull request merged**
